src/main/java/org/apache/commons/compress/archivers/zip/ParallelScatterZipCreator.java

```diff
@@ -17,49 +17,145 @@
+ * </PRE>
+ *
+ * or (better parallelization, writing compressed results on the fly)
+ *
+ * <PRE>
+ * zipCreator.addArchiveEntry(..)
+ * zipCreator.addArchiveEntry(..)
+ *
+ * ZipArchiveOutputStream zipOutputStream = ...
+ * zipCreator.startWriteTo(zipOutputStream);
+ *
+ * zipCreator.addArchiveEntry(..)
+ * zipCreator.addArchiveEntry(..)
+ *
+ * zipCreator.waitFinishWriteTo();
+ * </PRE>
  */
 public class ParallelScatterZipCreator {
-    private final List<ScatterZipOutputStream> streams = synchronizedList(new ArrayList<ScatterZipOutputStream>());
+
+    private static final ParallelScatterZipEntry EOF_MARKER = new ParallelScatterZipEntry(null, 0, null);
+
     private final ExecutorService es;
+    private final boolean closeExecutorService;
     private final ScatterGatherBackingStoreSupplier backingStoreSupplier;
-    private final List<Future<Object>> futures = new ArrayList<>();
 
-    private final long startedAt = System.currentTimeMillis();
-    private long compressionDoneAt = 0;
-    private long scatterDoneAt;
+    private final List<ThreadBackingStore> threadBackingStores = synchronizedList(new ArrayList<ThreadBackingStore>());
+
+    private BlockingQueue<Future<ParallelScatterZipEntry>> futureWriteQueue = new LinkedBlockingDeque<>();
+    private Future<?> mainWriter;
+    private Exception mainWriterEx;
+
+    private final AtomicLong compressionElapsed = new AtomicLong();
+    private final AtomicLong scatterElapsed = new AtomicLong();
+
+    public static class ParallelScatterZipEntry {
+        private final ZipArchiveEntry zipArchiveEntry;
+        private final long len;
+        private final ThreadBackingStore threadBackingStore;
+
+        ParallelScatterZipEntry(ZipArchiveEntry zipArchiveEntry,
+                long len, ThreadBackingStore threadBackingStore) {
+            this.zipArchiveEntry = zipArchiveEntry;
+            this.len = len;
+            this.threadBackingStore = threadBackingStore;
+        }
+
+        InputStream createInputStream() {
+            return new BoundedInputStream(threadBackingStore.inputStream, len);
+        }
+
+    }
+
+    /** inner input/output Streams backing store, per Thread */
+    private class ThreadBackingStore {
+        ScatterGatherBackingStore backingStore;
+        OutputStream ouputStream;
+        InputStream inputStream;
+
+        ThreadBackingStore(ScatterGatherBackingStore backingStore) throws IOException {
+            this.backingStore = backingStore;
+            this.ouputStream = createOutputStream();
+            this.inputStream = backingStore.getInputStream();
+        }
+
+        // ... adapter for non-existent method backingStore.createOutputStream();
+        // TODO move to class ScatterGatherBackingStore
+        OutputStream createOutputStream() {
+            return new OutputStream() {
+                @Override
+                public void write(int b) throws IOException {
+                    byte[] data = new byte[] { (byte) b };
```

Comment:
Allocating one-byte arrays? Won't this hurt performance?

Reply:
Hi,
This method is necessary to implements OutputStream methods, and delegate to 
   ScatterGatherBackingStore.writeOut(byte[] data, int offset, int length) throws IOException;
Anyway, this method from this private adapter class is not used.

Maybe, It would be better that ScatterGatherBackingStore expose directly an OutputStream (but it would change signatures), or expose a unitary "write(byte data)"

Regards,